source/Engine.cpp

```diff
@@ -1209,15 +1209,23 @@ void Engine::EnterSystem()
 		for(const System::FleetProbability &fleet : system->Fleets())
 			if(fleet.Get()->GetGovernment() && Random::Int(fleet.Period()) < 60)
 				fleet.Get()->Place(*system, newShips);
-		for(const System::HazardProbability &hazard : system->Hazards())
+
+		auto CreateWeather = [this](const Hazard::Probability &hazard, Point origin)
+		{
 			if(Random::Int(hazard.Period()) < 60)
 			{
 				const Hazard *weather = hazard.Get();
 				int hazardLifetime = weather->RandomDuration();
 				// Elapse this weather event by a random amount of time.
 				int elapsedLifetime = hazardLifetime - Random::Int(hazardLifetime + 1);
-				activeWeather.emplace_back(weather, hazardLifetime, elapsedLifetime, weather->RandomStrength());
+				activeWeather.emplace_back(weather, hazardLifetime, elapsedLifetime, weather->RandomStrength(), origin);
```

Comment:
I'm not sure we have implemented validity checks for hazards, but since they can be attached to a system by an event we should verify only fully defined hazards are added to the active lists.